### ISSUE TITLE
`CREATE TABLE` with 2048 partitions racks up 8 GB of memory on QD

### ISSUE THREAD

**Issue by d** [greenplum-db/gpdb#10947]
### Greenplum version or build
7.0 Alpha (master branch commit 1681ae162dcdf9e4 as of writing)

### autoconf options used ( config.status --config )
```
'--enable-orca' '--with-gssapi' '--enable-mapreduce' '--with-perl' '--with-libxml' '--with-python' '--disable-gpcloud' '--disable-pxf' '--enable-gpfdist' '--enable-depend' '--enable-debug' '--enable-debug-extensions' '--prefix=/build/install' '--with-includes=/build/install/include' '--with-libs=/build/install/lib' 'CXX=ccache c++' 'CC=ccache cc' 'CFLAGS=-O3 -fno-omit-frame-pointer' 'CXXFLAGS=-O3 -fno-omit-frame-pointer'

```
### Expected behavior
To be fair, I don't know what to expect. I understand the cost of keeping all the information to facilitate an `ABORT` later on is reasonable, but we might want to explicitly acknowledge this in the documentation.

### Actual behavior
```
psql: ERROR:  Canceling query because of high VMEM usage. Used: 7368MB, available 819MB, red zone: 7372MB (runaway_cleaner.c:200)
```

### Step to reproduce the behavior

```sql
CREATE TABLE foo(a int, b smallint, c int) DISTRIBUTED BY (a)
PARTITION BY RANGE (b)
(START(0) END(2048) EVERY(1))
;
```

### QD memory usage with `CREATE TABLE`
To be clear, I'm not suggesting this is worth fixing, because we should be switching to the new syntax anyway. But the quadratic memory usage concerns me a little:

 #parts | Greenplum 7 QD memory usage
---|---
512 | 541900800
576 | 683589632
640 | 841764864
704 | 1016619008
768 | 1208057856
832 | 1415794688
896 | 1640443904
960 | 1881153536
1024 | 2138632192
1536 | 4793049088
1664 | 5622087680
1792 | 6517141504
1920 | 7478210560
1936 | 7603052544

![chart](https://user-images.githubusercontent.com/440892/95277782-796c5900-0803-11eb-8812-587b38d76dab.png)

In contrast, the memory usage in Greenplum 6 appears to be linear:

#parts | bytes
-- | --
512 | 34795520
1024 | 67854336
1536 | 100442112
2048 | 133406720
2560 | 166010880
3072 | 198647808
3584 | 231378944
4096 | 264343552
8192 | 526233600
9216 | 591691776
10240 | 657137664
11264 | 722567168
12288 | 788295680
13312 | 853671936
14336 | 919162880
15360 | 984772608
16384 | 1051013120
32768 | 2053681152
65536 | 4106260480

![bytes vs  #parts](https://user-images.githubusercontent.com/440892/95349476-fafbcf80-0873-11eb-9e40-3c1db3129435.png)

### QD memory usage with a `psql` loop
The code suggests that `CREATE TABLE foo ... (PARTITION foo_0 ..., PARTITION foo_1 ...)` would have been equivalent to the following client side loop:
```sql
CREATE TABLE foo(a int, b smallint, c int)
PARTITION BY RANGE (b);

SET client_min_messages TO warning;
SELECT format('CREATE TABLE %s partition OF %s FOR VALUES FROM (%s) TO (%s)', "partition", root, i, i+1)
FROM (
    SELECT format('foo_%s', i) AS partition, 'foo' AS root, i
    FROM generate_series(0, :nparts - 1) i
) t; \gexec

```
Where `:nparts` is a `psql` variable that I can change from the command line.

So I tried:

#parts | bytes
-- | --
512 | 17162240
1024 | 52097024
1536 | 106209280
2048 | 180834304
2560 | 272113664
3072 | 385474560
3584 | 515633152
4096 | 665759744
6144 | 1460404224
8192 | 2565660672
10240 | 3973984256
12288 | 5692960768

So it is still quadratic, but it grows noticeably slower:
![chart (1)](https://user-images.githubusercontent.com/440892/95349074-89bc1c80-0873-11eb-9546-ec7489b0f70f.png)

**Comment by ashwinstar** [comment]
Any ideas what is the memory usage on GPDB6? if its regression on GPDB 7 due to syntax mapping from GPDB syntax to PostgreSQL 12, then we definitely need to look into it before GA, else if not regression then falls into general improvement of product category.

**Comment by d** [comment]
> Any ideas what is the memory usage on GPDB6? if its regression on GPDB 7 due to syntax mapping from GPDB syntax to PostgreSQL 12, then we definitely need to look into it before GA, else if not regression then falls into general improvement of product category.

Hi @ashwinstar I updated the post with some measurement:

1. Greenplum 7 using `CREATE TABLE` convenience syntax to define a bunch of partitions
1. Greenplum 6 using the same (the only in the version) convenience syntax
1. Greenplum 7 using upstream partition syntax with a `psql` loop

TL;DR
1. Both the upstream syntax and Greenplum syntax exhibit quadratic memory growth, which is theoretically tolerable.
1. However, the convenience syntax seems to be way more (by a large multiplicative constant factor) expensive memory-wise than the supposedly equivalent upstream syntax
1. As a frame of reference, the Greenplum 6 server exhibits a linear growth in memory consumption.

Uneducated comment: I haven't dug deep into the code, but my uneducated guess is it might be impossible to get back the linearity, but at least the raw memory consumption is a bit overblown.

**Comment by ashwinstar** [comment]
@d please can you post how you got the memory used on QD for the stmt to trace out the numbers. That would be helpful to investigate and fix this issue.

**Comment by JunfengYang** [comment]
@d Hi Jesse, I plan to explore the memory issue. How did you get the memory usage per `parts`? Is it from `TopMemoryContext`?

**Comment by JunfengYang** [comment]
After dig into the memory context, found two `MemoryContext` may have issues.
For the GPDB's syntax
```
CREATE TABLE foo(a int, b smallint, c int) DISTRIBUTED BY (a)
PARTITION BY RANGE (b)
(START(0) END(512) EVERY(1));
```
1. The `PortalContext` is extremely large.
`PortalContext: 556321960 total in 71988 blocks; 3678352 free (18988 chunks); 552643608 used:`
2. There are lots of `partition descriptor` memory context under `CacheMemoryContext`.
```
CacheMemoryContext: 2113600 total in 10 blocks; 271336 free (37 chunks); 1842264 used
    partition descriptor: 53528 total in 12 blocks; 6840 free (0 chunks); 46688 used: foo
      partition descriptor: 53512 total in 12 blocks; 6904 free (0 chunks); 46608 used: foo
        partition descriptor: 53480 total in 12 blocks; 6968 free (0 chunks); 46512 used: foo
          partition descriptor: 53464 total in 12 blocks; 7032 free (0 chunks); 46432 used: foo
            partition descriptor: 53432 total in 12 blocks; 7096 free (0 chunks); 46336 used: foo
              partition descriptor: 53416 total in 12 blocks; 7160 free (0 chunks); 46256 used: foo
                partition descriptor: 53384 total in 12 blocks; 7224 free (0 chunks); 46160 used: foo
                  partition descriptor: 53368 total in 12 blocks; 7288 free (0 chunks); 46080 used: foo
                    partition descriptor: 53336 total in 12 blocks; 7352 free (0 chunks); 45984 used: foo
                      partition descriptor: 53320 total in 12 blocks; 7416 free (0 chunks); 45904 used: foo
                        partition descriptor: 53288 total in 12 blocks; 7480 free (0 chunks); 45808 used: foo
......
```

And for pg's syntax which also creates 512 tables. 
**Note I add `BEGIN` and `END` here.**
```
BEGIN;
\set nparts 512

CREATE TABLE foo(a int, b smallint, c int)
PARTITION BY RANGE (b);

SET client_min_messages TO warning;
SELECT format('CREATE TABLE %s partition OF %s FOR VALUES FROM (%s) TO (%s)', "partition", root, i, i+1)
FROM (
    SELECT format('foo_%s', i) AS partition, 'foo' AS root, i
    FROM generate_series(0, :nparts - 1) i
) t; \gexec

END;
```
1. The `PortalContext` is much less than GPDB/s syntax.
`PortalContext: 2143352 total in 273 blocks; 21552 free (86 chunks); 2121800 used`
2. There also are lots of `partition descriptor` memory context under `CacheMemoryContext`.
```
CacheMemoryContext: 2113600 total in 10 blocks; 263120 free (35 chunks); 1850480 used
    partition descriptor: 53528 total in 12 blocks; 6840 free (0 chunks); 46688 used: foo
      partition descriptor: 53512 total in 12 blocks; 6904 free (0 chunks); 46608 used: foo
        partition descriptor: 53480 total in 12 blocks; 6968 free (0 chunks); 46512 used: foo
          partition descriptor: 53464 total in 12 blocks; 7032 free (0 chunks); 46432 used: foo
            partition descriptor: 53432 total in 12 blocks; 7096 free (0 chunks); 46336 used: foo
              partition descriptor: 53416 total in 12 blocks; 7160 free (0 chunks); 46256 used: foo
......
```
This is because the partition description(the number of partitions) catalog changes after create each partition table that causes
invalid the relation cache and rebuild the relation from the new catalog. So this is the right behavior.

Then the main issue is why we use a lot of memory in `PortalContext`.

Compare with GPDB6, it does use less memory but the partition logic is totally different since we don't have partition rules and other logic anymore. So there's no need to compare with GPDB6.

Then I just focus on why the new logic from upstream uses much memory which shouldn't.
After I dig into the cache invalid logic, it increases lots of memory usage once a cache invalid.
I finally find the root cause.

The reason is for each `RelationBuildDesc` call, it'll call `RelationBuildPartitionDesc` to generate partition desc.
And the function `RelationBuildPartitionDesc` has lots of memory leaks in the current memory context which is `PortalContext`.
Normally, PG's syntax's lifecycle of a `PortalContext` is much shorter than our GPDB's partition syntax. So this is not a big deal.
But in our case, it'll cause a large memory leak since we generate subpartition create statements in same portal.
And then I check the commit history, before this commit d3f48dfae42f9655425d1f58f396e495c7fb7812, the leaked memory actually lives in a short live memory context. But it gets refactored out and uses `CLOBBER_FREED_MEMORY` or `CLOBBER_CACHE_ALWAYS` to maintain the leaked memory when needed in `RelationBuildDesc`.

I try to build with `-DCLOBBER_CACHE_ALWAYS`, but I can not start the demo cluster.

**Comment by JunfengYang** [comment]
Manually set `#define RECOVER_RELATION_BUILD_MEMORY 1` to enable the temp context in `RelationBuildDesc `, the memory usage for the same statement reduces to 17M compare with more than 500M before.

**Comment by JunfengYang** [comment]
`CLOBBER_FREED_MEMORY` or `CLOBBER_CACHE_ALWAYS` are used for test purpose and pg don't have this kind of issue, so to resolve the high memory usage for partition table, we may bring back the temp memory context used in `RelationBuildPartitionDesc` (which get removed in d3f48dfae42f9655425d1f58f396e495c7fb7812), any suggestions?

**Comment by JunfengYang** [comment]
And also based on the upstream thread https://www.postgresql.org/message-id/15943.1552601288@sss.pgh.pa.us.

> > Having CLOBBER_CACHE_ALWAYS behave differently sounds horrible.

> I'm not entirely convinced of that.  It's already pretty different...

>> We maintain a free list of AllocSetContexts nowadays, so creating a 
>> short-lived context should be pretty cheap. Or if it's still too 
>> expensive, we could create one short-lived context as a child of 
>> TopMemoryContext, and reuse that on every call, resetting it at the end 
>> of the function.

>RelationBuildDesc potentially recurses, so a single context won't do.
>Perhaps you're right that the context freelist would make this cheap
>enough to not matter, but I'm not sure of that either.

>What I'm inclined to do to get the buildfarm pressure off is to set
>things up so that by default we do this only in CCA builds, but there's
>a #define you can set from the compile command line to override that
>decision in either direction.  Then, if somebody wants to investigate
>the performance effects in more detail, they could twiddle it easily.
>Depending on the results, we could alter the default policy.
>regards, tom lane

We may need to test whether `-DRECOVER_RELATION_BUILD_MEMORY=1` affects performance.

**Comment by d** [comment]
Amazing investigation, this is awesome @JunfengYang

**Comment by JunfengYang** [comment]
The fix merged into Master as commit: d553ec4a0042d13bb7cf2224a8b1b6595399170c.
So close the issue.
